src/Compilers/CSharp/Portable/Binder/Binder_Deconstruct.cs

```diff
@@ -0,0 +1,417 @@
+        {
+            var receiverSyntax = receiver.Syntax;
+
+            if (receiver.Type.IsDynamic())
+            {
+                Error(diagnostics, ErrorCode.ERR_CannotDeconstructDynamic, receiverSyntax);
+                outPlaceholders = default(ImmutableArray<BoundDeconstructValuePlaceholder>);
+
+                return BadExpression(receiverSyntax, receiver);
+            }
+
+            var analyzedArguments = AnalyzedArguments.GetInstance();
+            var outVars = ArrayBuilder<OutDeconstructVarPendingInference>.GetInstance(numCheckedVariables);
+            DiagnosticBag bag = null;
+
+            try
+            {
+                for (int i = 0; i < numCheckedVariables; i++)
+                {
+                    var variable = new OutDeconstructVarPendingInference(assignmentSyntax);
+                    analyzedArguments.Arguments.Add(variable);
+                    analyzedArguments.RefKinds.Add(RefKind.Out);
+                    outVars.Add(variable);
+                }
+
+                const string methodName = "Deconstruct";
+                var memberAccess = BindInstanceMemberAccess(
+                                        receiverSyntax, receiverSyntax, receiver, methodName, rightArity: 0,
+                                        typeArgumentsSyntax: default(SeparatedSyntaxList<TypeSyntax>), typeArguments: default(ImmutableArray<TypeSymbol>),
+                                        invoked: true, diagnostics: diagnostics);
+
+                memberAccess = CheckValue(memberAccess, BindValueKind.RValueOrMethodGroup, diagnostics);
+                memberAccess.WasCompilerGenerated = true;
+
+                if (memberAccess.Kind != BoundKind.MethodGroup)
+                {
+                    return MissingDeconstruct(receiver, assignmentSyntax, numCheckedVariables, diagnostics, out outPlaceholders, receiver);
+                }
+
+                // After the overload resolution completes, the last step is to coerce the arguments with inferred types.
+                // That step returns placeholder (of correct type) instead of the outVar nodes that were passed in as arguments.
+                // So the generated invocation expression will contain placeholders instead of those outVar nodes.
+                // Those placeholders are also recorded in the outVar for easy access below, by the `SetInferredType` call on the outVar nodes.
+                bag = DiagnosticBag.GetInstance();
+                BoundExpression result = BindMethodGroupInvocation(
+                                            receiverSyntax, receiverSyntax, methodName, (BoundMethodGroup)memberAccess, analyzedArguments, bag, queryClause: null,
+                                            allowUnexpandedForm: true);
+
+                result.WasCompilerGenerated = true;
+                diagnostics.AddRange(bag);
+
+                if (bag.HasAnyErrors())
```

Comment:
I missed this during the previous review. It might be inappropriate to use HasAnyErrors here. There are at least two reasons for that:
- It resolves lazy diagnostics, which could create circularity problems in certain contexts.
- Lazy errors shouldn't really fail the binding. An example would be an obsolete Deconstruct method. Yes, it is an error to use it, but that shouldn't prevent us from recording the actual binding information and continuing to the next level. I.e. we shouldn't complain that we couldn't find a suitable method. 

Consider using HasAnyResolvedErrors instead and adding a test with an obsolete Deconstruct method.